File: src/lib/scala/Slate.Api/src/main/scala/slate/core/apis/core/Help.scala
/**
<slate_header>
  url: www.slatekit.com
  git: www.github.com/code-helix/slatekit
  org: www.codehelix.co
  author: Kishore Reddy
  copyright: 2016 CodeHelix Solutions Inc.
  license: refer to website and/or github
  about: A Scala utility library, tool-kit and server backend.
  mantra: Simplicity above all else
</slate_header>
  */

package slate.core.apis.core

import slate.common.ListMap
import slate.core.apis.ApiContainer
import slate.core.apis.doc.{ApiVisitor, DocConsole, Doc}

class Help(val ctn:ApiContainer, lookup:ListMap[String, Apis], docs:Doc) {


  /**
   * handles help request for all the areas supported
   *
   * @return
   */
  def help():Unit = {
  }


  /**
   * handles help request for a specific area
   *
   * @param area
   * @return
   */
  def helpForArea(area:String):Unit =
  {
    // Guard
    require(lookup.contains(area), s"${area} not found")

    val doc = new DocConsole()
    val visitor = new ApiVisitor()
    visitor.visitApis(area, lookup(area), doc)
  }


  /**
   * handles help request for a specific api
   *
   * @param area
   * @param apiName
   * @return
   */
  def helpForApi(area:String, apiName:String):Unit =
  {
    // Guard
    require(lookup.contains(area), s"${area} not found")
    require( lookup(area).contains(apiName), s"${apiName} not found")

    val apis = lookup(area)
    val api = apis(apiName)
    val doc = new DocConsole()
    val visitor = new ApiVisitor()
    visitor.visitApi(api,apiName, doc)
  }


  /**
   * handles help request for a specific api action
   *
   * @param area
   * @param apiName
   * @param actionName
   * @return
   */
  def helpForAction(area:String, apiName:String, actionName:String):Unit =
  {
    // Guard
    require(lookup.contains(area), s"${area} not found")
    require( lookup(area).contains(apiName), s"${apiName} not found")
    require( lookup(area)(apiName).contains(actionName), s"${actionName} not found")

    val apis = lookup(area)
    val api = apis(apiName)
    val doc = new DocConsole()
    val visitor = new ApiVisitor()
    visitor.visitApiAction(api, apiName, actionName, doc)
  }
}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/core/Help.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis.core

import slatekit.apis.ApiServer
import slatekit.apis.tools.docs.ApiVisitOptions
import slatekit.apis.tools.docs.ApiVisitor
import slatekit.apis.tools.docs.Doc
import slatekit.apis.tools.docs.DocUtils
import slatekit.common.types.Content
import slatekit.common.requests.Request
import slatekit.results.*
import slatekit.results.builders.Outcomes


sealed class HelpType {
    object All    : HelpType()
    object Area   : HelpType()
    object Api    : HelpType()
    object Action : HelpType()
}

class Help(val host: ApiServer, val routes: Routes, val docKey: String?, val docBuilder: () -> Doc) {

    fun process(req: Request): Outcome<Content> {
        val result = DocUtils.isHelp(req)
        return when (result) {
            is Success -> build(req, result.value)
            is Failure -> result
        }
    }

    /**
     * Handles help request on any part of the api request. Api requests are typically in
     * the format "area.api.action" so you can type help on each part / region.
     * e.g.
     * 1. area ?
     * 2. area.api ?
     * 3. area.api.action ?
     */
    fun build(req: Request, helpType: HelpType): Outcome<Content> {
        return if (!DocUtils.hasDocKey(req, docKey ?: "")) {
            Outcomes.denied("Unauthorized access to API docs")
        } else {
            val content = when (helpType) {
                // 1: ? = help on all
                HelpType.All -> {
                    areas()
                }
                // 2: {area} ? = help on area
                HelpType.Area -> {
                    area(req.parts[0])
                }
                // 3. {area}.{api} = help on api
                HelpType.Api -> {
                    api(req.parts[0], req.parts[1])
                }
                // 4. {area}.{api}.{action} = help on api
                HelpType.Action -> {
                    action(req.parts[0], req.parts[1], req.parts[2])
                }
            }
            Outcomes.success(Content.html(content))
        }
    }

    /**
     * handles help request for all the areas supported
     *
     * @return
     */
    fun areas(): String {
        val doc = docBuilder()
        val visitor = ApiVisitor(routes)
        visitor.visitAreas(doc)
        return doc.toString()
    }

    /**
     * handles help request for a specific area
     *
     * @param area
     * @return
     */
    fun area(area: String): String {
        val doc = docBuilder()
        val visitor = ApiVisitor(routes)
        visitor.visitApis(area, doc)
        return doc.toString()
    }

    /**
     * handles help request for a specific api
     *
     * @param area
     * @param apiName
     * @return
     */
    fun api(area: String, api: String): String {
        val doc = docBuilder()
        val visitor = ApiVisitor(routes)
        visitor.visitActions(area, api, doc)
        return doc.toString()
    }

    /**
     * handles help request for a specific api action
     *
     * @param area
     * @param apiName
     * @param actionName
     * @return
     */
    fun action(area: String, name: String, action: String): String {
        val doc = docBuilder()
        val visitor = ApiVisitor(routes)
        val api = routes.api(area, name)
        api?.actions?.get(action)?.let { act ->
            visitor.visitAction(api, act, doc, true, options = ApiVisitOptions())
        }
        return doc.toString()
    }
}
